File: app/models/helpdesk.py
from __future__ import annotations
from datetime import datetime, timezone
from sqlalchemy import Enum, ForeignKey, Integer, String, Text, DateTime
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base
from app.models.common import TicketStatus

class HelpdeskQueue(Base):
    __tablename__ = "helpdesk_queues"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(64), unique=True, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    categories: Mapped[list["HelpdeskCategory"]] = relationship(back_populates="queue", cascade="all, delete-orphan")
    tickets: Mapped[list["SupportTicket"]] = relationship(back_populates="queue_ref")

class HelpdeskCategory(Base):
    __tablename__ = "helpdesk_categories"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    queue_id: Mapped[int] = mapped_column(ForeignKey("helpdesk_queues.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(64), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    queue: Mapped["HelpdeskQueue"] = relationship(back_populates="categories")
    tickets: Mapped[list["SupportTicket"]] = relationship(back_populates="category_ref")

class SupportTicket(Base):
    __tablename__ = "support_tickets"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    customer_id: Mapped[int | None] = mapped_column(ForeignKey("customers.id"), nullable=True)
    queue_id: Mapped[int | None] = mapped_column(ForeignKey("helpdesk_queues.id"), nullable=True)
    category_id: Mapped[int | None] = mapped_column(ForeignKey("helpdesk_categories.id"), nullable=True)
    assignee_id: Mapped[int | None] = mapped_column(ForeignKey("portal_users.id"), nullable=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[TicketStatus] = mapped_column(Enum(TicketStatus), default=TicketStatus.open, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc), nullable=False)

    customer: Mapped["Customer"] = relationship(back_populates="tickets")
    queue_ref: Mapped["HelpdeskQueue"] = relationship(back_populates="tickets")
    category_ref: Mapped["HelpdeskCategory | None"] = relationship(back_populates="tickets")
    assignee: Mapped["PortalUser | None"] = relationship(back_populates="assigned_tickets")

File: crm-portal-ts/src/models/helpdesk.ts
import { Column, Entity, JoinColumn, ManyToOne, OneToMany, PrimaryGeneratedColumn } from "typeorm";
import { TicketStatus } from "./common.js";
import { Customer } from "./customer.js";

@Entity("helpdesk_queues")
export class HelpdeskQueue {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "varchar", length: 64, unique: true })
    name!: string;

    @Column({ type: "text", nullable: true })
    description?: string;

    @Column({ type: "integer", name: "sort_order", default: 0 })
    sortOrder!: number;

    @OneToMany(() => HelpdeskCategory, (category) => category.queue)
    categories!: HelpdeskCategory[];

    @OneToMany(() => SupportTicket, (ticket) => ticket.queue)
    tickets!: SupportTicket[];
}

@Entity("helpdesk_categories")
export class HelpdeskCategory {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "queue_id" })
    queueId!: number;

    @Column({ type: "varchar", length: 64 })
    name!: string;

    @Column({ type: "text", nullable: true })
    description?: string;

    @ManyToOne(() => HelpdeskQueue, (queue) => queue.categories, { onDelete: "CASCADE" })
    @JoinColumn({ name: "queue_id" })
    queue!: HelpdeskQueue;

    @OneToMany(() => SupportTicket, (ticket) => ticket.category)
    tickets!: SupportTicket[];
}

@Entity("support_tickets")
export class SupportTicket {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "customer_id", nullable: true })
    customerId?: number;

    @Column({ type: "integer", name: "queue_id", nullable: true })
    queueId?: number;

    @Column({ type: "integer", name: "category_id", nullable: true })
    categoryId?: number;

    @Column({ type: "integer", name: "assignee_id", nullable: true })
    assigneeId?: number;

    @Column({ type: "varchar", length: 255 })
    title!: string;

    @Column({ type: "text" })
    body!: string;

    @Column({
        type: "simple-enum",
        enum: TicketStatus,
        default: TicketStatus.open,
    })
    status!: TicketStatus;

    @Column({ type: "datetime", name: "created_at", default: () => "CURRENT_TIMESTAMP" })
    createdAt!: string;

    @Column({ type: "datetime", name: "updated_at", default: () => "CURRENT_TIMESTAMP" })
    updatedAt!: string;

    @ManyToOne(() => Customer, { nullable: true, onDelete: "SET NULL" })
    @JoinColumn({ name: "customer_id" })
    customer?: Customer | null;

    @ManyToOne(() => HelpdeskQueue, (queue) => queue.tickets, { nullable: true, onDelete: "SET NULL" })
    @JoinColumn({ name: "queue_id" })
    queue?: HelpdeskQueue | null;

    @ManyToOne(() => HelpdeskCategory, (category) => category.tickets, { nullable: true, onDelete: "SET NULL" })
    @JoinColumn({ name: "category_id" })
    category?: HelpdeskCategory | null;
}
